### PR TITLE
feat: `unstack` receives kwarg `fillvalue`

### LINKED ISSUE #2698: control fill value for missing cells in `unstack`
Currently, when `unstack` encounters combinations of column and row keys that are not present in the DataFrame, the corresponding cell in the output is filled with `missing`.  It would be nice to be able to control the fill value used with a kwarg.  

For context, I hit this using `unstack` as a convenient way to make a confusion matrix (for which it's especially nice since you can add additional grouping variables on the rows), but cells that are missing really correspond to `0` in that situation.

### PR DESCRIPTION
**Pull request by pstorozenko**
Closes #2698 

I thought about checking every column separately for `promote_type`, but it would be a breaking change since right now value column with eltype `T` is always converted to `Vector{Union{Missing, T}}`. We may think about changing it in another PR.

### CODE REVIEW

**Comment by bkamins** [comment]
>  but it would be a breaking change

Why do you think so? I would assume that if `fillvalue` is `missing` which is the default and old behavior in your PR we would get the same (so this means that no old code would be broken - the result would stay the same).

File: NEWS.md [hunk 1]
```diff
@@ -18,6 +18,11 @@
   (notably `PooledArray` and `CategoricalArray`) or when they contained only
   integers in a small range.
   ([#2812](https://github.com/JuliaData/DataFrames.jl/pull/2812))
+* the `unstack` function receives new keyword argument `fillvalue`
+  (with `missing` default) that fills combination of not encountered
```
Comment (bkamins) [inline NEWS.md:22]:
```suggestion
  (with `missing` default) that is used to fill combinations of not encountered
```

File: NEWS.md [hunk 2]
```diff
@@ -18,6 +18,11 @@
...
+  rows and columns. This feature allows to distinguish between missings in
+  value column and just missing row/column combinations.
```
Comment (bkamins) [inline NEWS.md:24]: I would say that the main use case is `fillvalue=0` if you have count data.

File: test/reshape.jl [hunk 3]
```diff
@@ -654,4 +654,27 @@ end
...
+    dfu = unstack(df, :variable, :value)
+    @test dfu ≅ dfu1
+    @test eltype(dfu[!, :Var1]) == Union{Missing, Int}
+
+    delete!(df, 4)
+    for (sentinel, fcoltype) in zip([1, 1., "1", nothing], [Int, Float64, Any, Union{Int, Nothing}])
+        dfu_m = unstack(df, :variable, :value, fillvalue=sentinel)
+        @test dfu_m[2, :Var2] == sentinel
+        @test eltype(dfu_m[!, :Var2]) == fcoltype
+    end
```
Comment (bkamins) [inline test/reshape.jl:677]: It is very important here to add tests for `CategoricalVector`, `ChainedVector` and maybe also `PooledVector`.
The reason is to check if the line https://github.com/JuliaData/DataFrames.jl/pull/2828/files#diff-9b09ee01805b1e911ca9c715199a11e9c872da2dfb66250a5ae8097cf8e67216R429 will produce what you expect in different combinations of cases. E.g. if `CategoricalVector` will be retained given different types of `fillvalue`.

**Comment by bkamins** [comment]
> I thought about checking every column separately

Ah - after reviewing I see what you mean. What you have implemented is OK I think. One can run `identity.(df)` to do type narrowing later if needed.

However, it is important to test the additional scenarios I have mentioned I think. Especially for `CategoricalVector` check how `levels` behave in different combinations of inputs and in cases when original `CategoricalVector` is ordered/unordered.

Other than that - the PR looks good.

**Comment by bkamins** [comment]
@pstorozenko - merge conflicts require fixing + have you had time to add the tests I have suggested?

**Comment by bkamins** [comment]
@pstorozenko bump

**Comment by pstorozenko** [comment]
Sorry for not answering.
There are a few things I found when trying to create tests for Categorical Vector:

1. When dealing with vectors that are subtypes of `CategoricalArrays.AbstractCategoricalArray` we should create `unstacked_val` with `fillvalue` value and levels from `valuecol` + `fillvalue`.
2. When `fillvalue` is `missing` then it is not added to levels (that's how CategoricalArrays work AFAIK).

**Comment by bkamins** [comment]
Point 2 is correct.

Point 1 - there is a corner case when `fillvalue` is not allowed by `CategoricalArray`. As discussed - I will update the PR to check for corner cases. Thank you!

File: test/reshape.jl (hunk 3 above)
Comment (bkamins) [reply test/reshape.jl:694]: currently what we produce is disputable. I will comment above.

File: src/abstractdataframe/reshape.jl [hunk 4]
```diff
@@ -398,13 +421,15 @@ function _unstack(df::AbstractDataFrame, rowkeys::AbstractVector{Int},
...
     colref_map = df[col_group_row_idxs, colkey]
-
     if any(ismissing, colref_map) && !allowmissing
         throw(ArgumentError("Missing value in variable :$(_names(df)[colkey]). " *
                             "Pass `allowmissing=true` to skip missings."))
     end
+    unstacked_val = [fill!(similar(valuecol,
+                                   promote_type(eltype(valuecol), typeof(fillvalue)),
+                                   Nrow),
+                           fillvalue) for _ in 1:Ncol]
```
Comment (bkamins) [inline src/abstractdataframe/reshape.jl:431]: @nalimilan - this definition is problematic for categorical columns the problem is that:
```
julia> using CategoricalArrays

julia> valuecol = categorical(["a"])
1-element CategoricalArray{String,1,UInt32}:
 "a"

julia> fillvalue = ""
""

julia> Nrow = 1
1

julia> similar(valuecol, promote_type(eltype(valuecol), typeof(fillvalue)), Nrow)
1-element Vector{String}:
 #undef

julia> promote_type(eltype(valuecol), typeof(fillvalue))
String
```

Is there a generic idiom that would (without introducing dependency on CategoricalArrays.jl) allow us to take the union of types and if it is OK (like it should be in the example above) produce `CategoricalArray`? Or maybe you think we should change the definition of `similar` in CategoricalArrays.jl?

Comment (bkamins) [reply src/abstractdataframe/reshape.jl:431]: OK - I have thought about it. Given we currently allow:
```
julia> x = categorical([1,2,3])
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> CategoricalValue(3, x)
CategoricalValue{Int64, UInt32} 3
```
the behavior we have now should be OK.

File: test/reshape.jl [hunk 5]
```diff
@@ -654,6 +654,74 @@ end
...
+    df = DataFrame(factory=["Fac1", "Fac1", "Fac2"],
+                   variable=["Var1", "Var2", "Var1"],
+                   value=categorical([1, 2, 3]))
+    dfu = unstack(df, :variable, :value, fillvalue=CategoricalValue(0, categorical([0])))
+    @test dfu.Var1 == [1, 3]
+    @test typeof(dfu.Var1) <: CategoricalVector{Int}
+    @test dfu.Var2 ≅ [2, 0]
+    @test typeof(dfu.Var2) <: CategoricalVector{Int}
+    @test levels(dfu.Var1) == levels(dfu.Var2) == 0:3
+    dfu = unstack(df, :variable, :value, fillvalue=CategoricalValue("0", categorical(["0"])))
```
Comment (bkamins) [inline test/reshape.jl:717]: @nalimilan - this test in particular is something we should think of. My judgement is that it is OK to produce such an union.

**Review by bkamins — APPROVED** [review]

File: src/abstractdataframe/reshape.jl [hunk 6]
```diff
@@ -331,36 +335,55 @@ julia> unstack(long, :id, :variable, :value, renamecols=x->Symbol(:_, x))
    4 │     4       2.0       1.0       2.0
    5 │     5       2.0       1.0       3.0
    6 │     6       2.0       1.0       3.0
+
+julia> df = DataFrame(id= ["1", "1", "2"],
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl:339]:
```suggestion
julia> df = DataFrame(id=["1", "1", "2"],
```

File: src/abstractdataframe/reshape.jl (hunk 4 above)
Comment (nalimilan) [reply src/abstractdataframe/reshape.jl:431]: I think it's OK. If we wanted to change this we could adapt `promote_type` in CategoricalArrays, but it's orthogonal to this PR. There would certainly be advantages in promoting to `CategoricalValue`, though it could also create problems (e.g. when concatenating `Vector{Int} and `CategoricalArray{Union{String, Missing}}` the resulting type would be `Any` but CategoricalArrays doesn't handle that well).

File: NEWS.md [hunk 7]
```diff
@@ -22,6 +22,12 @@
   (notably `PooledArray` and `CategoricalArray`) or when they contained only
   integers in a small range.
   ([#2812](https://github.com/JuliaData/DataFrames.jl/pull/2812))
+* the `unstack` function receives new keyword argument `fillvalue`
```
Comment (nalimilan) [inline NEWS.md:25]: Maybe we should call this just `fill`? Or is that name useful in other contexts as a Boolean argument?

File: src/abstractdataframe/reshape.jl (hunk 4 above)
Comment (bkamins) [reply src/abstractdataframe/reshape.jl:431]: Yes - I think there would be more problems than benefits. We only need an easy way to create a "stand alone" `CategoricalValue`, as now it is `CategoricalValue(scalar, categorical([scalar]))` which is quite inconvenient.

File: NEWS.md (hunk 7 above)
Comment (bkamins) [reply NEWS.md:25]: No - this name is specific only to this function. I will rename it to `fill`.

**Comment by bkamins** [comment]
OK - I have renamed `fillvalue` to `fill`. The only potential problem would be if we used the `fill` function from Julia Base in the body of the function, but we do not use it. We only use `fill!` so all is OK.

**Comment by bkamins** [comment]
@nalimilan - any additional comments here?

File: src/abstractdataframe/reshape.jl [hunk 8]
```diff
@@ -229,6 +229,10 @@ Row and column keys will be ordered in the order of their first appearance.
 - `allowduplicates`: if `false` (the default) then an error an error will be thrown
   if combination of `rowkeys` and `colkey` contains duplicate entries; if `true`
   then  then the last encountered `value` will be retained.
+- `fill`: missing row/column combinations are filled with this value; The default
+  is `missing`. If the `value` column is a `CategoricalVector` and `fill`
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl:233]:
```suggestion
- `fill`: missing row/column combinations are filled with this value. The default
  is `missing`. If the `value` column is a `CategoricalVector` and `fill`
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!
